wallet/rpc: fix listdescriptors RPC fails to return descriptors with private key information when wallet contains descriptors missing any key

Eunovo opened the pull request:
_TLDR:
Currently, `listdescriptors [private=true]` will fail for a non-watch-only wallet if any descriptor has a missing private key(e.g `tr()`, `multi()`, etc.). This PR changes that while making sure `listdescriptors [private=true]` still fails if there no private keys. Closes #32078_ 

In non-watch-only wallets, it's possible to import descriptors as long as at least one private key is included. It's important that users can still view these descriptors when they need to create a backup—even if some private keys are missing ([#32078 (comment)](https://github.com/bitcoin/bitcoin/issues/32078#issuecomment-2781428475)). This change makes it possible to do so.

This change also helps prevent `listdescriptors true` from failing completely, because one descriptor is missing some private keys.

### Notes
- The new behaviour is applied to all descriptors including miniscript descriptors
- `listdescriptors true` still fails for watch-only wallets to preserve existing behaviour https://github.com/bitcoin/bitcoin/pull/24361#discussion_r920801352
- Wallet migration logic previously used `Descriptor::ToPrivateString()` to determine which descriptor was watchonly. This means that modifying the `ToPrivateString()` behaviour caused descriptors that were previously recognized as "watchonly" to be "non-watchonly". **In order to keep the scope of this PR limited to the RPC behaviour, this PR uses a different method to determine `watchonly` descriptors for the purpose of wallet migration.** A follow-up PR can be opened to update migration logic to exclude descriptors with some private keys from the `watchonly` migration wallet. 

### Relevant PRs
https://github.com/bitcoin/bitcoin/pull/24361
https://github.com/bitcoin/bitcoin/pull/32186

### Testing
Functional tests were added to test the new behaviour

EDIT
**`listdescriptors [private=true]` will still fail when there are no private keys because non-watchonly wallets must have private keys and calling `listdescriptors [private=true]` for watchonly wallet returns an error**

Eunovo committed:
walletrpc: reject listdes with priv key on w-only wallets

Eunovo committed:
test: Test listdescs with priv works even with missing priv keys

Co-authored-by: rkrux <rkrux.connect@gmail.com>

fanquake commented:
```bash
/Users/runner/work/bitcoin/bitcoin/src/test/fuzz/miniscript.cpp:1040:42: error: no matching member function for call to 'ToString'
    std::optional<std::string> str{node->ToString(parser_ctx)};
                                   ~~~~~~^~~~~~~~
/Users/runner/work/bitcoin/bitcoin/src/script/miniscript.h:830:17: note: candidate function template not viable: requires 2 arguments, but 1 was provided
    std::string ToString(const CTx& ctx, bool& has_priv_key) const {
                ^
/Users/runner/work/bitcoin/bitcoin/src/test/fuzz/miniscript.cpp:1248:31: error: no matching member function for call to 'ToString'
    const auto str2 = parsed->ToString(parser_ctx);
                      ~~~~~~~~^~~~~~~~
/Users/runner/work/bitcoin/bitcoin/src/script/miniscript.h:830:17: note: candidate function template not viable: requires 2 arguments, but 1 was provided
    std::string ToString(const CTx& ctx, bool& has_priv_key) const {
                ^
2 errors generated.
```

Eunovo committed:
wallet/migration: use HavePrivateKeys in place of ToPrivateString

ToPrivateString() behaviour will be modified in the following commits.

In order to keep the scope of this PR limited to the RPC behaviour,
this commit updates wallet migration to use 'Descriptor::HavePrivateKeys()'
in place of 'Descriptor::ToPrivateString()' to determine watchonly descriptors.

A follow-up PR can be opened to update migration logic to exclude
descriptors with some private keys from the watchonly migration wallet.

vicjuma commented:
This PR improves the usability of the `listdescriptors true` RPC, especially for descriptor-based (modern) wallets where some private keys may be missing.

However, it's unclear why someone would create a watch-only wallet and still choose to set `private=true`. Why not just use the default command, `bitcoin-cli listdescriptors` instead of `bitcoin-cli -named listdescriptors private=true`. The former already shows public keys? Unless the user is not certain about the nature of the descriptors they imported. I have used named arguments so that you can get the nature of the command `bitcoin-cli listdescriptors true`

achow101 commented:
Concept NACK

The purpose of the `private` argument is to show the private keys. If there are no private keys to show, I think it is a reasonable error to say that, rather than just showing the public descriptors. I think changing the behavior could result in a footgun where people use it thinking that they are being shown private keys when they actually are not.

vicjuma commented:
Concept NACK

It's unclear why someone should receive public keys with `private=true`. Why not just use the default command, `bitcoin-cli listdescriptors`, which already shows public keys? I assume the user is certain about the nature of the descriptors they have in their wallet. Maybe if I do not fully understand this PR.

rkrux commented:
>  If there are no private keys to show, I think it is a reasonable error to say that, rather than just showing the public descriptors.

I have not checked this PR yet but the original intention came from this issue: https://github.com/bitcoin/bitcoin/issues/32078#issuecomment-2781428475

The idea was to show results for descriptors that don't have _all_ the private keys but do have at least one. This is not for the case when the descriptor doesn't have _any_ private key.

I do like how the `importdescriptors` RPC returns a warning when the descriptor contains partial private keys: https://github.com/bitcoin/bitcoin/issues/32078#issuecomment-2783608466. Maybe a similar warning can be returned in this case for `listdescriptors` as well - to avoid the footgun.
```
[
  {
    "success": true,
    "warnings": [
      "Not all private keys provided. Some wallet functionality may return unexpected errors"
    ]
  }
]
```

Regardless of the above point, I do feel that preventing `listdescriptors true` from failing completely in the case a descriptor is missing few private keys is helpful for the user as the other descriptors with all the private keys will be returned in the response.
> This change also helps prevent listdescriptors true from failing completely, because one descriptor is missing some private keys.

vicjuma commented:
Maybe `-rpcwallet=<wallet>` will come in handy in this situation where you would have to use a different wallet for these situations to prevent the conflicts.

achow101 commented:
> The idea was to show results for descriptors that don't have all the private keys but do have at least one.

I think that's fine to show a result, but I don't think we should show any result if there are no private keys at all.

Eunovo commented:
> Concept NACK
> 
> The purpose of the `private` argument is to show the private keys. If there are no private keys to show, I think it is a reasonable error to say that, rather than just showing the public descriptors. I think changing the behavior could result in a footgun where people use it thinking that they are being shown private keys when they actually are not.

I think there has been a misunderstanding. **There are private keys to show.** This  PR doesn't change the behaviour of `listdescriptors true` RPC to show any result when there are no private keys. Non-watchonly wallet descriptors must have at least one private key. The RPC still fails for watch only wallets

Eunovo commented:
> I think that's fine to show a result, but I don't think we should show any result if there are no private keys at all.

No result is shown if there are no private keys

rkrux commented:
A similar problem happens in `gethdkeys private=true` as well and I think it's due to partial private keys when the following error is thrown:

```
➜  bitcoin git:(list-descriptors-with-partial-keys) ✗ bitcoinclireg -named gethdkeys private=true
error code: -1
error message:
map::at:  key not found
```

It appears to be coming from here when the `xprv` corresponding to the `xpub` can't be found: https://github.com/bitcoin/bitcoin/blob/7763e86afa045910a14ac9b2cd644927a447370b/src/wallet/rpc/wallet.cpp#L822

It's due to mismatch in the count of `wallet_xprvs` and `wallet_xprvs` calculated in the previous loop over `spkms`. 
I can create a separate issue later.

rkrux commented:
Concept ACK bb14e8ac7d38f3698d76ea3d6dce09e1387d59f3

Started reviewing the PR.

rkrux reviewed (commented):
Partial review bb14e8ac7d38f3698d76ea3d6dce09e1387d59f3

Eunovo commented on src/script/descriptor.cpp:
```diff
@@ -686,20 +698,23 @@ class DescriptorImpl : public Descriptor
         std::string ret;
-        ToStringHelper(nullptr, ret, compat_format ? StringType::COMPAT : StringType::PUBLIC);
+        bool has_priv_key;
+        ToStringHelper(nullptr, ret, compat_format ? StringType::COMPAT : StringType::PUBLIC, has_priv_key);
```
`ToStringHelper` now accepts a pointer to `has_priv_key` in https://github.com/bitcoin/bitcoin/pull/32471/commits/2a1f1f0e32829af43436eb4220164c1be86229bb
It does look cleaner, but there's a need to check for `nullptr` now.

Eunovo commented on src/script/miniscript.h:
```diff
@@ -840,10 +841,16 @@ struct Node {
         };
+        auto toString = [&ctx, &has_priv_key](Key key) -> std::string {
+            bool tmp{false};
+            auto key_str{ctx.ToString(key, tmp)};
```
Done

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -686,20 +698,23 @@ class DescriptorImpl : public Descriptor
     {
-        bool ret = ToStringHelper(&arg, out, StringType::PRIVATE);
+        bool has_priv_key;
+        assert(ToStringHelper(&arg, out, StringType::PRIVATE, has_priv_key));
```
I had something like this in mind because in all the descriptors that can be added in a non-watch-only wallet, there must be at least 1 private key that is available. So, it seemed reasonable to me to add an `Assume` here on `has_priv_key` being true. I don't suppose `ToPrivateString` function would be called in any case for a watch-only wallet.
```diff
diff --git a/src/script/descriptor.cpp b/src/script/descriptor.cpp
index bd85ddd9e2..b22bf3fa4a 100644
--- a/src/script/descriptor.cpp
+++ b/src/script/descriptor.cpp
@@ -710,6 +710,7 @@ public:
         // ToStringHelper should never fail for StringType::PRIVATE,
         // because it falls back to StringType::PUBLIC when no private ke
y is available.
         assert(ToStringHelper(&arg, out, StringType::PRIVATE, &has_priv_k
ey));
+        Assume(has_priv_key);
         out = AddChecksum(out);
         return has_priv_key;
     }
```

The `descriptor_tests` unit test did fail for one case when I used an `assert` instead, the functional tests though pass on my system. Maybe I have missed some use case due to which it fails. 
```bash
test/descriptor_tests.cpp:212: info: check parse_priv->ToPrivateString(keys_priv, prv1) has passed
test/descriptor_tests.cpp:216: info: check 'Private ser: combo(L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1)#p5326pcv Private desc: combo(L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1)' has passed
Assertion failed: (has_priv_key), function ToPrivateString, file descriptor.cpp, line 713.
unknown location:0: fatal error: in "descriptor_tests/descriptor_test": signal: SIGABRT (application abort requested)
test/descriptor_tests.cpp:218: last checkpoint
test/descriptor_tests.cpp:518: Leaving test case "descriptor_test"; testing time: 1795us
```

ISTM that the unit test case is more contrived than the functional tests as the below check fails.
```cpp
BOOST_CHECK(!parse_priv->ToPrivateString(keys_pub, prv1));
```

You can ignore if this doesn't make sense.

rkrux replied:
Sorry but I don't understand how the second statement follows the first one above. 
For a non-watch-only wallet, isn't it guaranteed that there must be at least one private key, which should turn `has_priv_key` to `true`?

rkrux replied:
> However, the descriptor unit tests often call ToPrivateString() on watch-only descriptors

Yeah, I don't know why this was done.

> To add this assertion, we will have to remove those checks, and I haven't seen the advantage in doing that.

Agree, doesn't need to be done right now.

Eunovo committed:
descriptors: add HavePrivateKeys()

Previously, to determine if a desc is watchonly, `ToPrivateString()`, was used.
It returns `false` if there is at least one pubkey in the descriptor for which
the provider  does not have a private key.

ToPrivateString() behaviour will change in the following commits to only
return `false` if no priv keys could be found for the pub keys in the descriptor.

HavePrivateKeys() is added here to replace the use of ToPrivateString() for determining
if a descriptor is 'watchonly'.

Co-authored-by: rkrux <rkrux.connect@gmail.com>

Eunovo commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -721,9 +721,27 @@ std::optional<MigrationData> LegacyDataSPKM::MigrateToDescriptor()
+            }
+        }
+        size_t pubkey_count{pubkeys.size() + ext_pubkeys.size()};
+        bool watchonly{privkey_count == 0 || pubkey_count > privkey_count};
```
Nice idea. Added you as Co-author in the relevant commit.

rkrux commented:
Thanks for addressing the comments, I will take one final look and then a-c-k.

Tagging @furszy for review as he noticed this issue as well few months back.

Sjors commented on src/script/descriptor.h:
```diff
@@ -111,6 +111,9 @@ struct Descriptor {
     /** Whether this descriptor will return one scriptPubKey or multiple (aka is or is not combo) */
     virtual bool IsSingleType() const = 0;
 
+    /** Whether this descriptor contains any private key material */
```
In 2c0a84f1e2ebe98bcf5066acb0de0ba5138c2101 _descriptors: add IsWatchOnly()_: this doesn't describe the actual behavior. IIUC:

```cpp
/** Whether this descriptor is missing any private key material */
```

Eunovo replied:
Thanks for this. I've updated the comment.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -626,6 +626,31 @@ class DescriptorImpl : public Descriptor
+            CKey key;
+            if (arg.GetKey(pubkey.GetID(), key)) {
+                privkey_count += 1;
+            } else if (output_type.has_value() && *output_type == OutputType::BECH32M && arg.GetKeyByXOnly(XOnlyPubKey(pubkey), key)) {
```
In https://github.com/bitcoin/bitcoin/commit/2c0a84f1e2ebe98bcf5066acb0de0ba5138c2101 descriptors: add IsWatchOnly(): since c++17 you can safely compare std::optional<T> with T, see item 21 here: https://en.cppreference.com/w/cpp/utility/optional/operator_cmp.html

I also find it easier to read if `privkey_count` and `pubkey_count` are tracked in the same manner, but feel free to ignore:

```diff
diff --git a/src/script/descriptor.cpp b/src/script/descriptor.cpp
index 98ae29ceb3..a86b4743b3 100644
--- a/src/script/descriptor.cpp
+++ b/src/script/descriptor.cpp
@@ -628,26 +628,27 @@ public:

     bool IsWatchOnly(const SigningProvider& arg) const override
     {
+        size_t pubkey_count{0};
         size_t privkey_count{0};
         std::set<CPubKey> pubkeys;
         std::set<CExtPubKey> ext_pubkeys;
         this->GetPubKeys(pubkeys, ext_pubkeys);
-        auto output_type{this->GetOutputType()};
-        for (auto pubkey : pubkeys) {
-            CKey key;
-            if (arg.GetKey(pubkey.GetID(), key)) {
+        for (const CPubKey pubkey : pubkeys) {
+            pubkey_count++;
+            CKey dummy;
+            if (arg.GetKey(pubkey.GetID(), dummy)) {
                 privkey_count += 1;
-            } else if (output_type.has_value() && *output_type == OutputType::BECH32M && arg.GetKeyByXOnly(XOnlyPubKey(pubkey), key)) {
+            } else if (this->GetOutputType() == OutputType::BECH32M && arg.GetKeyByXOnly(XOnlyPubKey(pubkey), dummy)) {
                 privkey_count += 1;
             }
         }
-        for (auto extpubkey : ext_pubkeys) {
+        for (const CExtPubKey extpubkey : ext_pubkeys) {
+            pubkey_count++;
             CKey dummy;
             if (arg.GetKey(extpubkey.pubkey.GetID(), dummy)) {
                 privkey_count += 1;
             }
         }
-        size_t pubkey_count{pubkeys.size() + ext_pubkeys.size()};
         return (privkey_count == 0 || pubkey_count > privkey_count);
     }
```

Sjors commented on src/test/miniscript_tests.cpp:
```diff
@@ -188,7 +188,7 @@ struct KeyConverter {
     }
 
-    std::optional<std::string> ToString(const Key& key) const {
+    std::string ToString(const Key& key, bool* has_priv_key = nullptr) const {
```
In b32c2b290f8e4dcb7fe54ee4c81714a3785920a7 _descriptor: ToPrivateString() pass if  at least 1 priv key exists_:

To preserve existing behaviour:

```cpp
if (has_priv_key) *has_priv_key = true;
```

It's a bit worrying that the test passes whether I set this to `false` or `true`. But that's a pre-existing issue, as the following change to the original code doesn't break the test either:

```diff
diff --git a/src/test/miniscript_tests.cpp b/src/test/miniscript_tests.cpp
index 0a32727f37..0989f0c50a 100644
--- a/src/test/miniscript_tests.cpp
+++ b/src/test/miniscript_tests.cpp
@@ -189,7 +189,7 @@ struct KeyConverter {
     }

     std::optional<std::string> ToString(const Key& key) const {
-        return HexStr(ToPKBytes(key));
+        return {};
     }

     miniscript::MiniscriptContext MsContext() const {
```

cc @darosior

Eunovo replied:
`has_priv_key` is only relevant when `ToPrivateString` is called on an actual Descriptor instance. IIUC, this test doesn't do that, so the value of `has_priv_key` does not matter here.
I'm not sure we should set it here, apart from the fact that it doesn't affect the test, the string returned doesn't have a private key.

Sjors reviewed (commented):
Mostly happy with d82dcf2eaa7efb3809ae559c8f97f74403a2ccd6, but in  b32c2b290f8e4dcb7fe54ee4c81714a3785920a7 _descriptor: ToPrivateString() pass if  at least 1 priv key exists_ I find the miniscript change hard to follow (mostly because I find the miniscript code itself hard to follow). Can you split that out in a seperate commit that doesn't change behaviour?

Tested with https://github.com/Sjors/bitcoin/pull/91.

Also left some code style suggestions.

Eunovo committed:
descriptor: ToPrivateString() pass if  at least 1 priv key exists

- Refactor Descriptor::ToPrivateString() to allow descriptors with
  missing private keys to be printed. Useful in descriptors with
  multiple keys e.g tr() etc.
- The existing behaviour of listdescriptors is preserved as much as
  possible, if no private keys are availablle ToPrivateString will
  return false

Eunovo committed:
descriptor: refactor ToPrivateString for providers

This commit modifies the Pubkey providers to return the public string
if private data is not available.
This is setup for a future commit to make Descriptor::ToPrivateString
return strings with missing private key information.

Co-authored-by: rkrux <rkrux.connect@gmail.com>

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -841,6 +841,38 @@ class DescriptorImpl : public Descriptor
+            CKey key;
+            if (arg.GetKey(pubkey.GetID(), key)) {
+                privkey_count += 1;
+            } else if (output_type.has_value() && output_type == OutputType::BECH32M && arg.GetKeyByXOnly(XOnlyPubKey(pubkey), key)) {
```
In 64213213b3691c9d93de95f880f820a640cd3c35 _descriptors: add IsWatchOnly()_: nit: you don't need `output_type.has_value()`

Sjors commented on src/test/fuzz/miniscript.cpp:
```diff
@@ -124,10 +124,13 @@ struct ParserContext {
     {
         auto it = TEST_DATA.dummy_key_idx_map.find(key);
-        if (it == TEST_DATA.dummy_key_idx_map.end()) return {};
+        if (it == TEST_DATA.dummy_key_idx_map.end()) {
```
In 49f41141c674f83d0b66262a5b21aad623704264 _descriptor: ToPrivateString() pass if  at least 1 priv key exists_: maybe for good measure:

```cpp
*has_priv_key = false;
```

In both places it seems better to check that `priv_key` exists, even if that's always the case in this fuzzer. IIUC in that case you can leave out `tmp{false}` and just pass `nullptr` below.

Sjors reviewed (approved):
ACK 45e06e05e7f8d638ecc91a1ba62e1b38a5133a2b

I find that the separation of 23431cde206bd6e18f274773f77bd914553bf2cc makes it easier to follow. The separation of https://github.com/bitcoin/bitcoin/commit/d2b84af58073630f87df32a7b785971fe32cb9be from https://github.com/bitcoin/bitcoin/commit/49f41141c674f83d0b66262a5b21aad623704264 is less useful, but fine by me.

@achow101 can you update your concept opinion based on the updated behaviour? I think it makes sense to return a result if _any_ private key is present in descriptor, rather than all.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -841,6 +841,38 @@ class DescriptorImpl : public Descriptor
         return true;
     }
 
+    bool IsWatchOnly(const SigningProvider& arg) const override
```
In 64213213b3691c9d93de95f880f820a640cd3c35 "descriptors: add IsWatchOnly()"

I think this function can be vastly simplified by calling `GetPrivKey` on all the `PubkeyProvider`, and `IsWatchOnly` on all subdescriptors.

nit: I don't like this function name since whether the descriptor is watchonly depends on the contents of the provided parameter. It's not a property like a `Is*` name implies. Watchonly also is not a property applied to descriptors as private keys are not actually part of the descriptor. Perhaps something like `HavePrivateKeys` instead.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -896,24 +896,27 @@ class DescriptorImpl : public Descriptor
 
     // NOLINTNEXTLINE(misc-no-recursion)
-    virtual bool ToStringSubScriptHelper(const SigningProvider* arg, std::string& ret, const StringType type, const DescriptorCache* cache = nullptr) const
+    virtual bool ToStringSubScriptHelper(const SigningProvider* arg, std::string& ret, const StringType type, bool* has_priv_key, const DescriptorCache* cache = nullptr) const
```
In 49f41141c674f83d0b66262a5b21aad623704264 "descriptor: ToPrivateString() pass if  at least 1 priv key exists"

`has_priv_key` should be a reference not a pointer. The pointer-ness is never used in this function.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -922,7 +925,8 @@ class DescriptorImpl : public Descriptor
                     break;
                 case StringType::PRIVATE:
-                    if (!pubkey->ToPrivateString(*arg, tmp)) return false;
+                    assert(has_priv_key != nullptr);
```
In 49f41141c674f83d0b66262a5b21aad623704264 "descriptor: ToPrivateString() pass if  at least 1 priv key exists"

If `has_priv_key` is never supposed to be null, then it should be a reference. Otherwise, this should be an `if`, we should not rely on caller behavior.

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -949,14 +955,17 @@ class DescriptorImpl : public Descriptor
+        bool has_priv_key{false};
+        // ToStringHelper should never fail for StringType::PRIVATE,
+        // because it falls back to StringType::PUBLIC when no private key is available.
+        assert(ToStringHelper(&arg, out, StringType::PRIVATE, &has_priv_key));
```
In 49f41141c674f83d0b66262a5b21aad623704264 "descriptor: ToPrivateString() pass if  at least 1 priv key exists"

`assert`s should not have side effects. If we are calling a function to actually do something, it should not be in an `assert`.

However, if the return value of `ToStringHelper` is always supposed to be `true`, then it should not be returning anything anyways. Furthermore, I don't see why `ToStringHelper` doesn't return whether the string has a private key, as we are doing in this function. That would allow us to drop `has_priv_key` from all of the function.

Eunovo replied:
I refactored the code as suggested and was able to drop `has_priv_key` from the parameter list. Now, `has_priv_key` is returned if `type == StringType::PRIVATE`.

achow101 reviewed (commented):
Concept ACK

Eunovo commented:
Thanks for the reviews! I have:
- renamed `IsWatchOnly()` to `HavePrivateKeys()`
- Added a test for `HavePrivateKeys()` to `descriptor_tests`
- Removed the unnecessary asserts from `ToStringHelper()` and `ToPrivateString()`
- Removed `has_priv_key` from `ToStringHelper()` and `ToSubStringHelper()` parameter lists

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -1418,24 +1425,34 @@ class TRDescriptor final : public DescriptorImpl
+        if (m_depths.empty()) {
+            // If there are no sub-descriptors and a PRIVATE string
+            // is requested, return `false` to indicate that the presence
+            // of a private key depends solely on the internal key, not on
+            // any sub-descriptor. This ensures correct behavior for
```
In deed439bff638abfc298a3b9c4f3d9f75fa18244 "descriptor: ToPrivateString() pass if  at least 1 priv key exists"

Micro nit if retouched

```diff
-            // of a private key depends solely on the internal key, not on
-            // any sub-descriptor. This ensures correct behavior for
+           // of a private key depends solely on the internal key (which is checked
+           // in the caller), not on any sub-descriptor. This ensures correct behavior for
```

Eunovo replied:
Done

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -2104,7 +2120,7 @@ struct KeyParser {
     }
 
-    std::optional<std::string> ToString(const Key& key) const
+    std::string ToString(const Key& key, bool* has_priv_key = nullptr) const
```
In https://github.com/bitcoin/bitcoin/commit/deed439bff638abfc298a3b9c4f3d9f75fa18244 "descriptor: ToPrivateString() pass if at least 1 priv key exists"

Nit: 
```diff
diff --git a/src/script/descriptor.cpp b/src/script/descriptor.cpp
index b6539ab12e..028a7299e5 100644
--- a/src/script/descriptor.cpp
+++ b/src/script/descriptor.cpp
@@ -2120,7 +2120,7 @@ struct KeyParser {
         return key;
     }
 
-    std::string ToString(const Key& key, bool* has_priv_key = nullptr) const
+    std::string ToString(const Key& key, bool*) const
     {
         return m_keys.at(key).at(0)->ToString();
     }
```

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -1524,11 +1541,12 @@ class StringMaker {
         std::string ret;
         if (m_private) {
-            if (!m_pubkeys[key]->ToPrivateString(*m_arg, ret)) return {};
+            assert(has_priv_key != nullptr);
```
In https://github.com/bitcoin/bitcoin/commit/deed439bff638abfc298a3b9c4f3d9f75fa18244 "descriptor: ToPrivateString() pass if at least 1 priv key exists"

Looks like the same argument applies here too?

https://github.com/bitcoin/bitcoin/pull/32471#discussion_r2258451488

> If has_priv_key is never supposed to be null, then it should be a reference.

rkrux reviewed (commented):
The miniscript code took some time for me to understand.

lgtm ACK 4241edfce6f9a4a51121b39bc20040c14d220551

rkrux reviewed (approved):
ACK 46860de809a1e9c7d0a1d716f77e9df7fb0bcd1c

```
git range-diff 4241edf...46860de
```

Sjors commented:
utACK 46860de809a1e9c7d0a1d716f77e9df7fb0bcd1c

I studied 4d07eff13002384ed59a3b7f592be56a25b88c15 from scratch since it had quite a few changes.

rkrux commented on src/script/descriptor.cpp:
```diff
@@ -885,13 +885,16 @@ class DescriptorImpl : public Descriptor
     virtual bool ToStringSubScriptHelper(const SigningProvider* arg, std::string& ret, const StringType type, const DescriptorCache* cache = nullptr) const
     {
         size_t pos = 0;
+        bool has_priv_key{false};
```
`has_priv_key` is used only for `PRIVATE` type, consider the following that limits its usage to avoid it being considered a misnomer - related unit/functional tests pass.

```diff
diff --git a/src/script/descriptor.cpp b/src/script/descriptor.cpp
index f29e73e2e6..ef977a5912 100644
--- a/src/script/descriptor.cpp
+++ b/src/script/descriptor.cpp
@@ -891,7 +891,7 @@ public:
             std::string tmp;
             bool subscript_res{scriptarg->ToStringHelper(arg, tmp, type, cache)};
             if (type != StringType::PRIVATE && !subscript_res) return false;
-            has_priv_key = has_priv_key || subscript_res;
+            if (type == StringType::PRIVATE) has_priv_key = has_priv_key || subscript_res;
             ret += tmp;
         }
         return type == StringType::PRIVATE ? has_priv_key : true;
```

Eunovo replied:
I refactored the code a bit and changed `has_priv_key` to `any_success`.

rkrux reviewed (changes requested):
Requesting changes because found an issue with the `musig()` descriptor that can be fixed in this PR itself.

Eunovo commented on src/test/descriptor_tests.cpp:
```diff
@@ -49,7 +49,7 @@ constexpr int SIGNABLE = 1 << 3; // We can sign with this descriptor (this is no
 constexpr int MIXED_PUBKEYS = 1 << 5;
 constexpr int XONLY_KEYS = 1 << 6; // X-only pubkeys are in use (and thus inferring/caching may swap parity of pubkeys/keyids)
-constexpr int MISSING_PRIVKEYS = 1 << 7; // Not all private keys are available, so ToPrivateString will fail.
+constexpr int MISSING_PRIVKEYS = 1 << 7; // Not all private keys are available, so ToPrivateString will fail and HavePrivateKeys() will return `true`.
```
I have updated the comment.

Eunovo commented on src/wallet/rpc/backup.cpp:
```diff
@@ -490,6 +491,9 @@ RPCHelpMan listdescriptors()
 
     const bool priv = !request.params[0].isNull() && request.params[0].get_bool();
+    if (wallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) && priv) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Can't get private descriptor string for watch-only wallets");
```
I don't believe the suggested message is an improvement, so I will leave it as is.

rkrux reviewed (approved):
lgtm ACK 8b4d4b3

Thanks for quickly addressing the comments.

```zsh
git range-diff 74521dd...8b4d4b3
```

I have tried to think if the fix covers all the cases that could possibly arise in various descriptors containing private keys partially. Essentially, the `ToPrivateString` function of the descriptor classes returns `true` even if there is a single private key available for it, whereas the newly added `HavePrivateKeys` function returns `true` when all the private keys for the descriptor are available & hence, it's used in the wallet migration code now. 
Very good that the latest push also tests for the missing private keys case in the `else` block in the descriptor unit tests!

Note: I have limited exposure to the `Miniscript` code as of now; I might re-review later that portion again, but looks ok right now.

This issue has popped up in my own local quite a few times where it became hard for me to continue my workflow, hope it gets checked in soon!

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/32471/commits/2b16014dfbe7e249e4929beafb4ce26415206e7a

rkrux reviewed (approved):
reACK 2b16014

```
git range-diff 8b4d4b3...2b16014
```

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -883,13 +883,17 @@ class DescriptorImpl : public Descriptor
     {
         size_t pos = 0;
+        bool is_private{type == StringType::PRIVATE};
+        bool any_success{!is_private};
```
In 23f8b6035ff6538134be3d3df8a3ab8ba47f86c0 _descriptor: ToPrivateString() pass if  at least 1 priv key exists_: the variable name `any_success` had me confused. If you have to retouch, it would be good to document its meaning or rename it.

```cpp
// If a public string is requested, fail if no subscript returns
a public string. If a  private string is requested, fail if no 
subscript returns a private string.
```

Seems like `any_success` was suggested here: https://github.com/bitcoin/bitcoin/pull/32471#discussion_r2342296606

Eunovo replied:
I added comments where `any_success` was used.

Sjors reviewed (commented):
Was in the middle of a review, but had to leave. Some comments.

Sjors commented on src/wallet/test/walletload_tests.cpp:
```diff
@@ -26,6 +26,7 @@ class DummyDescriptor final : public Descriptor {
     bool IsRange() const override { return false; }
     bool IsSolvable() const override { return false; }
     bool IsSingleType() const override { return true; }
+    bool HavePrivateKeys(const SigningProvider&) const override { return true; }
```
In 1441145f34f3cf6f9e30599a67defd276c9ef114 _descriptors: add HavePrivateKeys()_ nit: this is an odd default for the test, because e.g. `ToPrivateString` is `false` by default.

Eunovo replied:
Changed to `false`.

Sjors reviewed (commented):
I still need to (re-)review the non-test changes of 4dc37d89de0b8b39b3e5ca802067663cfc3bb529, but everything else looks fine.

I wrote an expansion of the fuzzer so it actually covers `has_priv_key`, see https://github.com/Eunovo/bitcoin/pull/3, but it can wait for a followup.

The commit message for 583797341106abab8e6e9bd4bbaf91d5cc7f0cc9 _walletrpc: reject listdes with priv key on w-only wallets_ could mention that this improves the error message.

Sjors commented:
ACK 9c7e4771b13d4729fd20ea08b7e2e3209b134fff

achow101 commented:
ACK 9c7e4771b13d4729fd20ea08b7e2e3209b134fff

w0xlt commented on test/functional/wallet_listdescriptors.py:
```diff
@@ -146,6 +146,36 @@ def run_test(self):
+        assert_equal(len(actual_descs), len(expected_descs))
+        for desc in actual_descs:
+            if desc not in expected_descs:
+                raise AssertionError(f"{desc} missing")
```
```suggestion
                raise AssertionError(f"Unexpected descriptor: {desc}")
```

w0xlt commented on src/script/descriptor.cpp:
```diff
@@ -1542,11 +1596,15 @@ class MiniscriptDescriptor final : public DescriptorImpl
+        auto res = m_node->ToString(StringMaker(arg, m_pubkey_args, type, cache), has_priv_key);
+        if (res) out = *res;
+        if (type == StringType::PRIVATE) {
+            Assume(res.has_value());
```
nit: if the `Assume()` somehow fails in release, out will be empty string rather than undefined.

```suggestion                                                               
            if (!res) out.clear();  
            Assume(res.has_value());
```

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/32471/commits/9c7e4771b13d4729fd20ea08b7e2e3209b134fff with minor nits

rkrux reviewed (approved):
re-ACK 9c7e4771b13d4729fd20ea08b7e2e3209b134fff

but with a caveat that I have only looked at the range-diff and the latest commit messages this time. I've reviewed this PR before quite a few times and have now developed review fatigue for this PR. I'm relying on my previous acks to bolster this reack. 
```
git range-diff 2b16014...9c7e477
```
